src/defs.hpp

```diff
@@ -56,21 +60,211 @@ struct LogicalLocation { // aggregate and POD type
   // These values can exceed the range of std::int32_t even if the root grid has only a
   // single MeshBlock if >30 levels of AMR are used, since the corresponding max index =
   // 1*2^31 > INT_MAX = 2^31 -1 for most 32-bit signed integer type impelementations
-  std::int64_t lx1, lx2, lx3;
-  int level;
+
+ private:
+  std::int64_t lx1_, lx2_, lx3_;
+  MortonNumber morton_;
+  int level_;
+
+ public:
+  LogicalLocation(int lev, std::int64_t l1, std::int64_t l2, std::int64_t l3)
+      : lx1_(l1), lx2_(l2), lx3_(l3), level_(lev), morton_(lev, l1, l2, l3) {}
+  LogicalLocation() : LogicalLocation(0, 0, 0, 0) {}
+
+  const auto &lx1() const { return lx1_; }
+  const auto &lx2() const { return lx2_; }
+  const auto &lx3() const { return lx3_; }
+  const auto &level() const { return level_; }
+  const auto &morton() const { return morton_; }
 
   // operators useful for sorting
   bool operator==(LogicalLocation &ll) {
-    return ((ll.level == level) && (ll.lx1 == lx1) && (ll.lx2 == lx2) && (ll.lx3 == lx3));
+    return ((ll.level() == level_) && (ll.lx1() == lx1_) && (ll.lx2() == lx2_) &&
+            (ll.lx3() == lx3_));
   }
   static bool Lesser(const LogicalLocation &left, const LogicalLocation &right) {
-    return left.level < right.level;
+    return left.level() < right.level();
   }
   static bool Greater(const LogicalLocation &left, const LogicalLocation &right) {
-    return left.level > right.level;
+    return left.level() > right.level();
+  }
+
+  bool IsContainedIn(const LogicalLocation &container) const {
+    if (container.level() > level_) return false;
+    const std::int64_t shifted_lx1 = lx1_ >> (level_ - container.level());
+    const std::int64_t shifted_lx2 = lx2_ >> (level_ - container.level());
+    const std::int64_t shifted_lx3 = lx3_ >> (level_ - container.level());
+    return (shifted_lx1 == container.lx1()) && (shifted_lx2 == container.lx2()) &&
+           (shifted_lx3 == container.lx3());
+  }
```

Comment:
Really nice to have this kind of check (and friends below) in the `LogicalLocation` itself. I feel like that's a lot cleaner as far as logic goes.